src/Env/EnvFactory.php

```diff
@@ -24,13 +25,36 @@
  */
 class EnvFactory
 {
-    public static function createEnv(EnvEnum $type, AppName $appName, \Traversable $dependencies): Env
+    public static function createEnv(EnvEnum $name, AppName $appName, \Traversable $dependencies): Env
     {
         return new Env(
-            (string) $type,
-            new Vagrant($type, $appName, VagrantBoxVersionResolver::resolve($dependencies)),
-            new Ansible($type, ...$dependencies),
-            new Make($type)
+            (string) $name,
+            new Vagrant($name, $appName, VagrantBoxVersionResolver::resolve($dependencies)),
+            new Ansible($name, ...$dependencies),
+            new Make($name)
         );
     }
+
+    public static function createEnvFromMetadata(string $rawName, array $metadata): Env
+    {
+        $configs = [];
+        $name = EnvEnum::create($rawName);
+        $hydrator = new VariableHydrator();
+
+        foreach ($metadata as $configClass => $perClassVars) {
+            $hydratedVars = [];
+
+            foreach ($perClassVars as $class => $vars) {
+                foreach ($vars as $data) {
+                    $var = (new \ReflectionClass($class))->newInstanceWithoutConstructor();
+                    $hydrator->hydrate($var, $data);
+                    $hydratedVars[] = $var;
+                }
+            }
+
+            $configs[] = new $configClass($name, ...$hydratedVars);
```

Comment:
Should probably define a named constructor instead, otherwise there is nothing guaranteeing the constructor looks like this, right ? (even if we own that code, I think it's a clearer path)